x-pack/metricbeat/module/sql/query/query.go

```diff
@@ -179,11 +205,95 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 	return nil
 }
 
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+// It calls m.fetchTableMode() or m.fetchVariableMode() depending on the response
+// format of the query.
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	db, err := sql.NewDBClient(m.Config.Driver, m.HostData().URI, m.Logger())
+	if err != nil {
+		return fmt.Errorf("cannot open connection: %w", err)
+	}
+	defer db.Close()
+
+	queries := m.Config.Queries
+	if len(queries) == 0 {
+		one_query := query{Query: m.Config.Query, ResponseFormat: m.Config.ResponseFormat}
+		queries = append(queries, one_query)
+	}
+
+	if !m.Config.FetchFromAllDatabases {
+		err = m.fetch(ctx, db, reporter, queries)
+		if err != nil {
+			m.Logger().Warn("error while fetching:", err)
+		}
+		return nil
+	}
+
+	// NOTE: Only mssql driver is supported for now because:
+	//
+	// * Difference in queries to fetch the name of the databases
+	// * The statement to select a named database (for subsequent statements
+	// to be executed) may not be generic i.e, USE statement (e.g., USE <db_name>)
+	// works for MSSQL but not Oracle.
+	//
+	// TODO: Add the feature for other drivers when need arises.
+	validQuery := queryDBNames(m.Config.Driver)
+	if validQuery == "" {
+		return fmt.Errorf("fetch from all databases feature is not supported for driver: %s", m.Config.Driver)
+	}
+
+	// Discover all databases in the server and execute given queries on each
+	// of the databases.
+	dbNames, err := db.FetchTableMode(ctx, queryDBNames(m.Config.Driver))
+	if err != nil {
+		return fmt.Errorf("cannot fetch database names: %w", err)
+	}
+
+	if len(dbNames) == 0 {
+		return errors.New("no database names found")
+	}
+
+	qs := make([]query, 0, len(queries))
+
+	for i := range dbNames {
+		// Create a copy of the queries as query would be modified on every
+		// iteration.
+		qs = qs[:0]                 // empty slice
+		qs = append(qs, queries...) // copy queries
+
+		val, err := dbNames[i].GetValue("name")
+		if err != nil {
+			m.Logger().Warn("error with database name:", err)
+			continue
+		}
+		dbName, ok := val.(string)
+		if !ok {
+			m.Logger().Warn("error with database name's type")
+			continue
+		}
+
+		// Prefix dbSelector to the query based on the driver
+		// provided.
+		// Example: USE <dbName>; @command (or @query)
+		for i := range qs {
+			qs[i].Query = dbSelector(m.Config.Driver, dbName) + " " + qs[i].Query
+		}
+
+		err = m.fetch(ctx, db, reporter, qs)
```

Comment:
Kind of pedantic comment: the underlying `Fetch()` implementation returns a bool that will be false in cases where the beat is shutting down. For instances where we're calling `Fetch()` in a loop, we should carry the false value up and break from the loop.

Reply:
Yeah, good point. But let me just re-confirm; are you talking about `reporter.Done`? Ref: https://github.com/elastic/beats/blob/1f4051f6f7153463499debf3dd51109d5a82a052/metricbeat/mb/mb.go#L175

Reply:
Or checking the bool returned by `reporter.Event`?

Reply:
Okay. Both `Event` and `Done` are dependent on `done` to make the caller of the methods stop.

Here: https://github.com/elastic/beats/blob/1f4051f6f7153463499debf3dd51109d5a82a052/metricbeat/mb/module/wrapper.go#L357

I'll see what's suits best in the code.

Reply:
Hey @fearful-symmetry,  please take a look. I've addressed your review comment \o/